pkg/controller/schedulingmanager/controller.go

```diff
@@ -43,37 +42,55 @@ type SchedulerController struct {
 
 	worker util.ReconcileWorker
 
-	scheduler map[string]schedulingtypes.Scheduler
+	schedulers map[string]*SchedulerWrapper
 
 	config *util.ControllerConfig
+}
 
-	runningPlugins sets.String
-	// mapping qualifiedname to template kind for managing plugins in scheduler
-	federatedKindMap map[string]string
+type SchedulerWrapper struct {
+	// To signal shutdown of scheduler and any associated routine.
+	stopChan chan struct{}
+	// Mapping qualifiedname to template kind for managing plugins in scheduler.
+	// This is needed because typeconfig could be of any name and we run plugins
+	// by federated kinds (eg FederatedDeployment). This also avoids running multiple
+	// plugins in case multiple typeconfigs are created for same federated kind.
+	pluginMap *util.SafeMap
```

Comment:
we're using 2 level locks here: a global `sync.RWMutex` for all (`schedulers`) and a local `util.SafeMap` lock for `pluginMap`.  It seems a little complex. Could you explain more about values to make this lock to `pluginMap`

Reply:
The general recommendation if using locks is to lock only for the scope needed and make that scope as small as possible. The lock in pluginMap is map specific and locks the reads and writes to it; The lock scope is minimal and thus locking overhead is minimal too. The safe map is recommended for usage in rest of the code too to avoid races.

The controller local mutex is mainly to protect rest of the data in reconcile from races (`SchedulingManager.schedulers` and `SchedulerWrapper.stopChan`). I could have used this same mutex to protect `pluginMap` too but chose otherwise as it makes one variable less in the overhead of maintaining all read/write accesses whenever we change code next.